7233 Egen komponent for valg av bestemmelse i årsavregning

fredrik-oh opened the pull request:
- Flyttet bestemmelse-select ut av medlemskapsperiodeSkjema og opprettet ny komponent BestemmelseSelect
 - Bestemmelse kan kun endres for behandlinger uten delt grunnlag
  - Henter trygdedekninger automatisk ved endring av bestemmelse
  - Justerte yup schema med validering av data innenfor medlemskapsperioder
  - Hvis vi kun har 1 gyldig trygdedekning settes denne automatisk og feltet blir readOnly
  - Erstattet forrigeMedlemskapsperioder ref med React state variabel lagredeMedlemskapsperioder som defaulter til medlemskapsperioder fra initiellData
- Slettet initiellBeregning, skal ikke være nødvendig lengre
- Alle kall til handleBeregnTrygdeavgiftsperioder henter nå formValues med getValues kall i stedet for bruk av watch

Commit:
7233 TOGGLE Trekk ut bestemmelse fra medlemskapsperiodeSkjema.tsx

wip: Trukket ut bestemmelse i egen select felles for alle medlemskapsperioder i aarsavregningForm.tsx

tbd: låsing av bestemmelse select for delt grunnlag, fetching av mulige dekninger i aarsavregningForm.tsx

Commit:
Merge branch 'master' into MELOSYS-7233_Bestemmelse_select_utenfor_medlemskapsperioder

# Conflicts:
#	src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx

Commit:
fikset feil etter merge

Commit:
Fjernet validering på bestemmelse i medlemskapsperioder
La til validering på toppnivå for bestemmelse
Fjernet oppdatering av medlemskapsperioder ved valg av bestemmelse

tbd: fetching av trygdedekninger i aarsavregningForm.tsx, generell ryddejobb etter funksjonalitet er på plass

Commit:
Flyttet henting av trygdedekninger fra medlemskapsperiodeSkjema.tsx til aarsavregningForm.tsx

tbd: generell ryddejobb etter funksjonalitet er på plass

Commit:
fikset en bug ved initiell innlasting av inntektskilder.tsx, la til kommentarer i aarsavregningUtenEllerDeltGrunnlag.tsx

tbd:
nullstilling av inntektskilder
sletting av lagrede medlemskapsperioder ved endring av bestemmelse
generell ryddejobb etter funksjonalitet er på plass

Commit:
Fjernet ubrukt variabel lagredeMedlemskapsperioder fra initiellData prop i aarsavregningForm.tsx

tbd:
nullstilling av inntektskilder
sletting av lagrede medlemskapsperioder ved endring av bestemmelse
generell ryddejobb etter funksjonalitet er på plass

Commit:
Henter trygdedekninger initielt i aarsavregningUtenEllerDeltGrunnlag.tsx
La til en changeHandler for endring av bestemmelse som sletter lagrede medlemskapsperioder og nullstiller nødvendige skjemafelt

tbd:
trekke ut bestemmelse select i egen komponent
generell ryddejobb etter funksjonalitet er på plass

Commit:
la til ny komponent bestemmelseSelect.tsx

tbd:
Fikse sletting, fungerer greit for uten grunnlag, men delt grunnlag skal ikke slette perioder fra grunnlag
generell ryddejobb etter funksjonalitet er på plass

Commit:
ryddet litt i bestemmelseSelect.tsx

Commit:
fjernet en kommentar som ikke skulle med :)

Commit:
fikset riktige default values i mapTilInntektskilderProps

Commit:
Fjernet emptyFieldDisabled som ikke skulle vært lagt til i inntektskilder
Fjernet redux dispatch, årsavregning får spesialtilpasset oppførsel for visning av medlemskapsperiode så denne er overflødig

tbd:
For bestemmelser med kun 1 gyldig dekning skal denne settes automatisk og feltet skal være readOnly
Inntektskilder fungerer ikke helt som forventet etter endring av bestemmelse

Commit:
For bestemmelser med kun 1 gyldig dekning skal denne settes automatisk og feltet skal være readOnly

tbd:
Ekstra lagringskall på ugyldige verdier grunnet avvik mellom verdier fra useWatch og det som valideres mot
Inntektskilder fungerer ikke helt som forventet etter endring av bestemmelse

Commit:
Stenger ned useEffects inntil endring av bestemmelse er ferdig med å oppdatere og evt lagre nye medlemskapsperioder.

Dette scenarioet oppstår dersom man endrer til en bestemmelse som kun har en gyldig dekning.

Commit:
Ryddet litt i aarsavregningUtenEllerDeltGrunnlag.lastInitiellData.
Kaster en feil dersom vi treffer på scenarioet der årsavregningen har flere innvilgede perioder med ulik bestemmelse (Skal ikke skje)

Commit:
Merge branch 'master' into MELOSYS-7233_Bestemmelse_select_utenfor_medlemskapsperioder

# Conflicts:
#	src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningForm.tsx
#	src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx

Commit:
fjernet en kommentar som ikke skulle med

Commit:
fikset noe gjenglemt rot i aarsavregningUtenEllerDeltGrunnlagSchema.jsx

Commit:
fjernet en kommentar og en uønsket eslint warning

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlag.tsx:
```diff
@@ -175,7 +177,7 @@ export function AarsavregningUtenEllerDeltGrunnlag({ bekreft, oppdaterStatus, ha
           );
         } else {
-          // Vanlig innlastning. Når det ikke er delt grunnlag initielt.
+          // Vanlig innlastning. Delt og uten grunnlag
```
Det er vel ikke "delt grunnlag" her? Det er alllerede sjekk på "harDeltGrunnlag" på forrige punkt.

fredrik-oh replied:
Ikke så lett å se med mye kode, men jo den if blokken håndterer oppretting og mapping av medlemskapsperioder for delt grunnlag første gang det lastes inn.

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlag.tsx:
```diff
@@ -184,10 +186,36 @@ export function AarsavregningUtenEllerDeltGrunnlag({ bekreft, oppdaterStatus, ha
 
         const erInitiellMappingForDeltGrunnlag = harDeltGrunnlag && aarsavregningRes && !aarsavregningRes.nyttGrunnlag;
 
+        let bestemmelse = "";
```
Kan vi flytte den koden inn i egen funksjon? Denne useEffecten er utrolig svær nå, så jeg tror det hadde vært nyttig å få delt slikt opp i forklarlige funksjonsnavn.

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlag.tsx:
```diff
@@ -184,10 +186,36 @@ export function AarsavregningUtenEllerDeltGrunnlag({ bekreft, oppdaterStatus, ha
+            bestemmelse = førsteBestemmelse;
+          } else {
+            throw new Error(
+              "Kan ikke laste inn årsavregning fordi grunnlag eller behandlingsresultat har innvilgede medlemskapsperioder med ulik bestemmelse",
```
Jeg tror ikke denne sjekken er nødvendig. Det er valideringer på lagring, og det er validering på FTRL-flyten. Man kan med sikkerhet si at alle bestemmelser er helt like. Dersom man likevel er usikker og vil ha dette, ville jeg kanskje heller ha utvidet behandlingsresultatDto i backend, og lagt på bestemmelse der (bestemmelse ligger på Behandlingsresultat som ett felt)

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/bestemmelseSelect.tsx:
```diff
@@ -0,0 +1,124 @@
+  );
+
+  return (
+    <Forms.Select
```
skulle du ikke gjøre bredden fit-content?

fredrik-oh replied:
nah, overlater styling til andre når vi begynner å nærme oss prodsetting

fredrik-oh replied:
Har litt tid til overs så kan like så godt fikse styling :)

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx:
```diff
@@ -96,35 +91,16 @@ export function MedlemskapsperiodeSkjema({
-              ))}
-            </Forms.Select>
-          </Nav.Column>
           <Nav.Column className="trygdedekning">
```
Bredden på trygdedekning virker å være litt for smal nå. Mulig det var sånn før også, men når dekning er låst, så får man ikke sett  hele "Full dekning". Ikke kritisk, men dette bør være relativt enkelt å fikse, selv om dette ikke er en design-PR

fredrik-oh replied:
nah, overlater styling til andre når vi begynner å nærme oss prodsetting

fredrik-oh replied:
Har litt tid til overs så kan like så godt fikse styling :)

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningForm.tsx:
```diff
@@ -322,8 +332,8 @@ export function AarsavregningForm({
   useEffect(() => {
-    const lagreMedlemskapsperioder = async () => {
-      if (redigerbart) {
+    const lagreMedlemskapsperioderEffect = async () => {
```
Har ikke undersøkt lenge hvor feilen er, men det er en syklusproblem i en useEffect et eller annet sted. Det blir evig loop med lagring av medlemskapsperioder. Hvis du legger til to medlemskapsperioder i uten grunnlag, endrer tom-dato, så ser du effekten.

fredrik-oh replied:
Fikset

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningForm.tsx:
```diff
@@ -322,8 +332,8 @@ export function AarsavregningForm({
   useEffect(() => {
-    const lagreMedlemskapsperioder = async () => {
-      if (redigerbart) {
+    const lagreMedlemskapsperioderEffect = async () => {
```
Når man legger til ny medlemskapsperiode, blir det to ulike POST-request, med samme data. Det skjedde iallefall etter å oppleve uendelig PUT-loop.

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/bestemmelseSelect.tsx:
```diff
@@ -0,0 +1,124 @@
+      control={control}
+      readOnly={!redigerbart || harDeltGrunnlag}
+      onChange={(valgtBestemmelse) => {
+        if (valgtBestemmelse && valgtBestemmelse !== bestemmelse) {
```
Blir change kalt hvis ikke  man har endret bestemmelse da? Barae lurer på om denne if'en er unødvendig

fredrik-oh replied:
react hook form vil ikke reagere på endringen, men vår onChange ville gjort det uten sjekken

fredrik-oh replied:
fikset

IsaSalemHame commented on src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/bestemmelseSelect.tsx:
```diff
@@ -0,0 +1,124 @@
+        const trygdedekningerResponse = await Api.LovligeKombinasjoner.hentTrygdedekninger(nyBestemmelse);
+        setTrygdedekninger(trygdedekningerResponse);
+
+        if (medlemskapsperioder.every((periode: Medlemskapsperiode) => periode.id === ULAGRET_MEDLEMSKAPSPERIODE_ID)) {
```
Du har på en måte samme if sjekk lengre nede. Er det nødvendig med begge? Denne burde være plenty, om denne også er nødvendig i det hele tatt.

fredrik-oh replied:
Når jeg tenker meg om burde ikke denne sjekken stanse alt som skjer under uansett. Trygdedekning i medlemskapsperiode og inntektskilde++ i inntektskilder burde uansett nullstilles. Justerer så vi ikke sender et unødvendig slett kall dersom vi ikke har lagrede perioder, men gjør nullstilling av form

Commit:
trekker ut deler av lastInitiellData i egne metoder

Commit:
La til flag for å styre lagring av medlemskapsperioder. Fikser inf loop av lagringskall ved rask endring av dato frem og tilbake.

Renamet en variabel i aarsavregningForm.tsx

Commit:
Fjernet faktisk buggy if sjekk i handleBestemmelseChange. Burde uansett nullstille skjemaverdier i medlemskapsperioder og inntektskilder selv om periodene ikke er lagret.
Fjernet unødvendig if check fra onChange prop i bestemmelseSelect.tsx

Commit:
Erstattet forrigeMedlemskapsperioder ref med react state variabel lagredeMedlemskapsperioder. Defaulter til medlemskapsperioder fra initiellData og oppdaterer denne kun etter at lagring har skjedd.
Fjernet sjekk på length fra medlemskapsperioderHarBrukerEndringer
Fjernet initiellBeregning logikk, klarer ikke å se at det er nødvendig og tester ok etter sletting.

Commit:
Stenger skjemaet ved endring av bestemmelse siden denne operasjonen tar litt tid, kan fjernes raskt ved å reverte skjemaErRedigerbart

Commit:
ny todo kommentar

IsaSalemHame reviewed (approved):
Bra jobbet! Blitt en del ryddigere nå, og lettere å følge med, særlig med 
å kvitte oss med "prevMedlemskapsperioder" osv. Har fire kommentarer som vi må følge opp, men som ikke trengs å tas her.

- Slett-knappen trigger ikke ny validering. Man kan for eksempel la det føre til oppholdsperioder, så vises ikke felmeldingen. Kommenterer i Canvas.
- Beregningen skjer for tidlig, slik at det fører til feil slik som at inntektsperioder som egentlig ikke burde finnes sendes til backend. Dette fikses i valideringsoppgave.
-- Blitt enig om at du legger til en TODO her.
- Opplevde én gang `Kan ikke beregne trygdeavgift uten medlemskapsperioder`, men dette er usikkert om jeg har gjort noe rart eller ikke. Har prøvd å reprodusere uten hell. Sier ifra om at vi må være oppmerksom på dette i test.
- Medlemskapsperioder rerendres for mye, for hver endring. Egen oppgave.
